internal/fs/kernel_list_cache_test.go

```diff
@@ -0,0 +1,353 @@
+// Copyright 2024 Google Inc. All Rights Reserved.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+// A collection of tests which tests the kernel-list-cache feature, in which
+// directory listing 2nd time is served from kernel page-cache unless not invalidated.
+// Base of all the tests: how to detect if directory listing is served from page-cache
+// or from GCSFuse?
+// (a) GCSFuse file-system ensures different content, when listing happens on the same directory.
+// (b) If two consecutive directory listing for the same directory are same, that means
+//     2nd listing is served from kernel-page-cache.
+// (c) If not then, both 1st and 2nd listing are served from GCSFuse filesystem.
+
+package fs_test
+
+import (
+	"os"
+	"path"
+	"testing"
+	"time"
+
+	"github.com/googlecloudplatform/gcsfuse/v2/internal/config"
+	"github.com/stretchr/testify/assert"
+	"github.com/stretchr/testify/suite"
+)
+
+const (
+	kernelListCacheTtlSeconds = 1000
+)
+
+type KernelListCacheTestCommon struct {
+	suite.Suite
+	fsTest
+}
+
+func (t *KernelListCacheTestCommon) SetupSuite() {
+	t.serverCfg.ImplicitDirectories = true
+	t.serverCfg.MountConfig = &config.MountConfig{
+		ListConfig: config.ListConfig{
+			KernelListCacheTtlSeconds: kernelListCacheTtlSeconds,
+		}}
+	t.serverCfg.RenameDirLimit = 10
+	t.fsTest.SetUpTestSuite()
+}
+
+func (t *KernelListCacheTestCommon) SetupTest() {
+	t.createFilesAndDirStructureInBucket()
+	cacheClock.SetTime(time.Date(2015, 4, 5, 2, 15, 0, 0, time.Local))
+}
+
+func (t *KernelListCacheTestCommon) TearDownTest() {
```

Comment:
I think we should delete the files directories created by SetupTest here.
There are error logs like the following on running tests:
```
{"timestamp":{"seconds":1719562345,"nanos":424749656},"severity":"ERROR","message":"fuse_errors: *fuseops.RmDirOp error: directory not empty"}
{"timestamp":{"seconds":1719562345,"nanos":430420365},"severity":"ERROR","message":"fuse_errors: *fuseops.RmDirOp error: directory not empty"}
```